### ISSUE TITLE
WAMP Specification: Router initiated shutdown

### ISSUE THREAD

**Issue by davidchappelle** [tavendo/WAMP#127]
# Observations

Based on the basic profile specifications the protocol requires both peers to send a goodbye message to each other when a session is being closed. So if a component initiates the close:

    component === goodbye ===> router ===> goodbye ===> component

If the router initiates the close:

    router === goodbye ===> component ===> goodbye ===> router

My question is, in the case of a router shutting down, how is the anticipated goodbye response from the component handled? The router would basically have to block on shutdown until all of the goodbye responses from the components have been received. However, I do not see a code path for this in crossbar.io. Are the goodbye responses from the components simply ignored in this case?

# Proposed Specification Amendment

With a router initiated close of a single session, the router should wait and timeout the receipt of peer goodbye. It should also perform and timeout transport closing handshake, if the WAMP transport in use has that (eg websocket).

With a router shutting down normally, it should send goodbye but probably not wait on receiving goodbye reply from all sessions, for the reasons you outline. Same holds true for closing handshake of underlying transport. However, it should not hard drop the transports (eg TCP)

With a router shutting down in emergency, it should just hard drop the underlying transports without any closing handshakes at the WAMP or transport level.

Nothing of this is mentioned in the spec, but it should.

**Comment by ecorm** [comment]
With router-initiated shutdowns, I think it's too costly for the router to wait for all peers to respond to GOODBYE messages. Instead, the router should send an ABORT with `wamp.close.system_shutdown` as the URI, which is currently permitted in the spec under https://wamp-proto.org/wamp_latest_ietf.html#name-basic-profile-uris 

> The Peer is shutting down completely - used as a `GOODBYE` (or `ABORT`) reason.
>
>    `wamp.close.system_shutdown`

**Comment by oberstet** [comment]
yes, agreed: sending ABORT and shutting down the socket router side after that sounds right, and also fits David's intention (I think)
